File: python/138. Copy List with Random Pointer.py
# Definition for singly-linked list with a random pointer.
# class RandomListNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return None
        table = {}
        table[id(None)] = None
        ptr = head
        while ptr:
            table[id(ptr)] = RandomListNode(ptr.label)
            ptr = ptr.next
        ptr = head
        while ptr:
            node = table[id(ptr)]
            node.next = table[id(ptr.next)]
            node.random = table[id(ptr.random)]
            ptr = ptr.next
        return table[id(head)]
        
        
            
            
        

File: cpp/101-150/138. Copy List with Random Pointer.cc
/**
 * Definition for singly-linked list with a random pointer.
 * struct RandomListNode {
 *     int label;
 *     RandomListNode *next, *random;
 *     RandomListNode(int x) : label(x), next(NULL), random(NULL) {}
 * };
 */
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
    	if(!head)
    		return NULL;
    	RandomListNode *newhead = new RandomListNode(0);
    	map<RandomListNode *, RandomListNode *> dict;
    	RandomListNode *ptr = head;
    	RandomListNode *newptr = newhead;
    	while(ptr){
    		newptr->next = new RandomListNode(ptr->label);
    		newptr->next->random = ptr->random;
    		dict[ptr] = newptr->next;
    		newptr = newptr->next;
    		ptr = ptr->next;
    	}
    	newptr = newhead->next;
    	while(newptr){
    		newptr->random = dict[newptr->random];
    		newptr = newptr->next;
    	}
    	RandomListNode *res = newhead->next;
    	delete newhead;
    	return res;
    }
};